src/mon/ConnectionTracker.cc

```diff
@@ -203,46 +203,51 @@ void ConnectionTracker::notify_rank_removed(int rank_removed)
   ldout(cct, 20) << "my rank before: " << rank << dendl;
 
   encoding.clear();
-  size_t starting_size = my_reports.current.size();
-  // erase the removed rank from everywhere
-  my_reports.current.erase(rank_removed);
-  my_reports.history.erase(rank_removed);
+  size_t starting_size = peer_reports.size();
+  auto pi = peer_reports.upper_bound(rank_removed);
+  // Remove the target rank and adjust everything that comes after.
   peer_reports.erase(rank_removed);
-  // Move ranks > rank_removed down by 1
-  // First in my_reports' history+current
-  auto ci = my_reports.current.upper_bound(rank_removed);
-  auto hi = my_reports.history.upper_bound(rank_removed);
-  while (ci != my_reports.current.end()) {
-    ceph_assert(ci->first == hi->first);
-    my_reports.current[ci->first - 1] = ci->second;
-    my_reports.history[hi->first - 1] = hi->second;
-    my_reports.current.erase(ci++);
-    my_reports.history.erase(hi++);
+  while (pi != peer_reports.end()) {
+    peer_reports[pi->first - 1] = pi->second; // copy content of next rank to ourself.
+    --peer_reports[pi->first - 1].rank; // update the rank of our new content.
+    peer_reports.erase(pi++); // destroy our next rank and move on.
   }
-  ceph_assert((my_reports.current.size() == starting_size) ||
-	      (my_reports.current.size() + 1 == starting_size));
 
-  // now move ranks down one in peer_reports
-  starting_size = peer_reports.size();
-  auto pi = peer_reports.upper_bound(rank_removed);
-  while (pi != peer_reports.end()) {
-    peer_reports[pi->first - 1] = pi->second;
-    peer_reports.erase(pi++);
+  // Go through everyone, remove & adjust current and history map.
+  for (auto pi = peer_reports.begin(); pi != peer_reports.end(); pi++) {
+    peer_reports[pi->first].current.erase(rank_removed);
+    peer_reports[pi->first].history.erase(rank_removed);
+    auto ci = peer_reports[pi->first].current.upper_bound(rank_removed);
+    auto hi = peer_reports[pi->first].history.upper_bound(rank_removed);
+    size_t starting_size_current = peer_reports[pi->first].current.size();
+    while (ci != peer_reports[pi->first].current.end()) {
+        ceph_assert(ci->first == hi->first);
+        peer_reports[pi->first].current[ci->first - 1] = ci->second;
+        peer_reports[pi->first].history[hi->first - 1] = hi->second;
+        peer_reports[pi->first].current.erase(ci++);
+        peer_reports[pi->first].history.erase(hi++);
+    }
```

Comment:
So I really, really don't like this. The peer_reports mechanism is designed so only the monitor generating a report updates the numbers, and then we assimilate them. Directly manipulating them like this is icky, and actually can lead to differing election outcomes during an upgrade from the old-way monitors to this way.

From the commit message, I think this is the real problem:
> However, there is a case where if we remove the highest rank of the quorum, then we will leave a the higest rank in report.current and report.history

If we just resolve that, which sounds like a fairly simple bug in ConnectionTracker::receive_peer_reports, I don't think we need to do this twiddling.

Reply:
I think even with the old code, we are already directly manipulating peer_reports by adjusting ranks: https://github.com/ceph/ceph/blob/main/src/mon/ConnectionTracker.cc#L171-L176.
my thoughts are that since we are already adjusting the ranks of peer_reports, might as well adjust the ranks of peer_reports.current and peer_reports.history as well so that they are aligned with each other.

I understand your concerns about manipulating them, but if we go with that, then we should just only remove and not adjust the ranks in peer_reports at all correct?

Reply:
> I think even with the old code, we are already directly manipulating peer_reports by adjusting ranks: https://github.com/ceph/ceph/blob/main/src/mon/ConnectionTracker.cc#L171-L176. my thoughts are that since we are already adjusting the ranks of peer_reports, might as well adjust the ranks of peer_reports.current and peer_reports.history as well so that they are aligned with each other.
> 
> I understand your concerns about manipulating them, but if we go with that, then we should just only remove and not adjust the ranks in peer_reports at all correct?

That link is just removing scores, not updating the individual reports?

Reply:
The old code moves the peer_report down so in a way it is adjusting the ranks implicitly since functions like ``get_total_connection_score`` rely on the position of the peer_report rather than the actual peer_report.rank.

If we use the old code we'll have this in this scenario:

```
we are in mon.0 point of view.

notify_rank_removed: 2

peer_reports before remove 2: {
0=rank=0, current links: {1=1, 2=1, 3=1, 4=1}, history: {1=0.9, 2=0.9 3=0.9, 4=0.9},
1=rank=1, current links: {0=1, 2=1, 3=1, 4=1}, history: {0=0.9, 2=0.9, 3=0.9, 4=0.9},
2=rank=2, current links: {0=1, 1=1, 3=1, 4=1}, history: {0=0.9, 2=0.9, 3=0.9, 4=0.9}, 
3=rank=3, current links: {0=1, 1=1, 2=1, 4=1}, history: {0=0.8, 1=0.7, 2=0.6, 4=0.9},
4=rank=4, current links: {0=1, 1=1, 2=1, 3=1}, history: {0=0.9, 1=0.6, 2=0.5, 3=0.7},
}

peer_reports after remove 2: {
0=rank=0, current links: {1=1, 2=1, 3=1, 4=1}, history: {1=0.9, 2=0.9 3=0.9, 4=0.9},
1=rank=1, current links: {0=1, 2=1, 3=1, 4=1}, history: {0=0.9, 2=0.9, 3=0.9, 4=0.9}, 
2=rank=3, current links: {0=1, 1=1, 2=1, 4=1}, history: {0=0.8, 1=0.7, 2=0.6, 4=0.9},
3=rank=4, current links: {0=1, 1=1, 2=1, 3=1}, history: {0=0.9, 1=0.6, 2=0.5, 3=0.7},
}

```

So now, rank 4 becomes rank 3, and rank 3 becomes rank 2.

let's say mon.0 call get_total_connection_score peer 3,  peer_report[2] would give us nothing, but it should have given us 0.9.

So what I am trying to say is that since we are moving the ranks down by 1 already with the old code. I don't think it is fair to not move the ranks in current and history.

Reply:
Moreover, the fact that the old code doesn't adjust the actual peer_report.rank leaves us a vulnerability as well since let's say after we remove ranks with the old code and we ping another monitor and they happened to to `assimilate_connection_reports` by us, then with our unadjusted peer_report.rank, we will populate the incorrect information on their peer_reports and this also can lead to differing election outcomes.

Reply:
@gregsfortytwo PTAL